### ISSUE TITLE
[Feature Request] Support DHCPv6

### ISSUE THREAD

**Issue by leoniDEV** [TechnitiumSoftware/DnsServer#265]
The usage of IPv6 is increasing and many ISP now provide IPv6 to their customers so will be useful to have also support for DHCPv6

**Comment by ShreyasZare** [comment]
Thanks for the feature request. Will add this in the to-do list. It may take some time though since there are many features already in pipeline.

**Comment by CyberMainstay** [comment]
Technitium is almost the perfect solution for my homelab (as well as SMB).

All it is lacking is DHCP6, and more DHCP options.

**Comment by ShreyasZare** [comment]
> Technitium is almost the perfect solution for my homelab (as well as SMB).
> 
> All it is lacking is DHCP6, and more DHCP options.

Thanks for the post. What DHCP options are you looking for? There is plan to implement a few options in next couple of releases so any specific option that is required can be accommodated.

**Comment by CyberMainstay** [comment]
> > Technitium is almost the perfect solution for my homelab (as well as SMB).
> > All it is lacking is DHCP6, and more DHCP options.
> 
> Thanks for the post. What DHCP options are you looking for? There is plan to implement a few options in next couple of releases so any specific option that is required can be accommodated.

Of course, anytime.  These are the options that I can see as very important:
- 4 - time server
- 60 - vendor type/ID
- 66 - tftp (PXE booting)
- 67 - filename (specify file for PXE booting)
- 82 - WLAN Controller
- 138 - discovery services (used by some WiFi Access points so they can find their WLAN controller - TP-Link Omada, UniFi, etc.)

**Comment by ShreyasZare** [comment]
> These are the options that I can see as very important:
> 
>   * 4 - time server
> 
>   * 60 - vendor type/ID
> 
>   * 66 - tftp (PXE booting)
> 
>   * 67 - filename (specify file for PXE booting)
> 
>   * 82 - WLAN Controller
> 
>   * 138 - discovery services (used by some WiFi Access points so they can find their WLAN controller - TP-Link Omada, UniFi, etc.)

First 4 options are already available. Will try to add the other 2.

**Comment by CyberMainstay** [comment]
> > These are the options that I can see as very important:
> > 
> > * 4 - time server
> > * 60 - vendor type/ID
> > * 66 - tftp (PXE booting)
> > * 67 - filename (specify file for PXE booting)
> > * 82 - WLAN Controller
> > * 138 - discovery services (used by some WiFi Access points so they can find their WLAN controller - TP-Link Omada, UniFi, etc.)
> 
> First 4 options are already available. Will try to add the other 2.

Awesome!  Thanks!

**Comment by ShreyasZare** [comment]
>  * 82 - WLAN Controller
> 
>  * 138 - discovery services (used by some WiFi Access points so they can find their WLAN controller - TP-Link Omada, UniFi, etc.)

@CyberMainstay can you confirm the options above? 

- Option 82 is [Relay Agent Information](https://www.rfc-editor.org/rfc/rfc3046.html), nothing about WLAN.
- Option 138 is [CAPWAP Access Controller addresses](https://www.rfc-editor.org/rfc/rfc5417.html).

You might want to refer to this list of options [here](https://www.iana.org/assignments/bootp-dhcp-parameters/bootp-dhcp-parameters.xhtml).

**Comment by CyberMainstay** [comment]
> > * 82 - WLAN Controller
> > * 138 - discovery services (used by some WiFi Access points so they can find their WLAN controller - TP-Link Omada, UniFi, etc.)
> 
> @CyberMainstay can you confirm the options above?
> 
>     * Option 82 is [Relay Agent Information](https://www.rfc-editor.org/rfc/rfc3046.html), nothing about WLAN.
> 
>     * Option 138 is [CAPWAP Access Controller addresses](https://www.rfc-editor.org/rfc/rfc5417.html).
> 
> 
> You might want to refer to this list of options [here](https://www.iana.org/assignments/bootp-dhcp-parameters/bootp-dhcp-parameters.xhtml).

Option 138 is specifically used by TP-Link Omada to help APs find the WLAN controller.  For some reason, my mind is drawing a blank on why I suggest option 82, but I've run across it before with another WLAN controller vendor that used it.

**Comment by ShreyasZare** [comment]
> Option 138 is specifically used by TP-Link Omada to help APs find the WLAN controller. For some reason, my mind is drawing a blank on why I suggest option 82, but I've run across it before with another WLAN controller vendor that used it.

Thanks for the confirmation. Will get the option 138 added. 

Option 82 seems to be too specialized that is supposed to be used by ISPs to communicate between the DHCP server and DHCP Relay Agents so does not seem to be useful for this project.

**Comment by BazlTech** [comment]
Any update on implementing DHCP6?

**Comment by ShreyasZare** [comment]
> Any update on implementing DHCP6?

The next update will include syslog support and the update after that is planned for DHCPv6. So it could be 3-4 months from now depending on how it goes.

**Comment by BazlTech** [comment]
> > Any update on implementing DHCP6?
> 
> The next update will include syslog support and the update after that is planned for DHCPv6. So it could be 3-4 months from now depending on how it goes.

That is awesome to hear!  I'll be greatly looking forward to trying it out.  DNS resolution to IPv6 releases will be a very welcomed addition!

**Comment by josemanimala** [comment]
Heya Folks,
Wanted to ask how the progress on implementing ipv6 dhcp support is going.

**Comment by josemanimala** [comment]
> > Heya @ShreyasZare, Wanted to ask how the progress on implementing ipv6 dhcp support is going.
> 
> Thanks for asking. It is getting a bit delayed due to time constrains since a few months. So, it may take some more time.

Heya @ShreyasZare,
Just checking in again to ask if there was progress on ipv6. 
Also, just curious on if you accept funding to prioritize features?

**Comment by ShreyasZare** [comment]
> Heya @ShreyasZare, Just checking in again to ask if there was progress on ipv6. 

@josemanimala Thanks for asking. The DNS server clustering feature is now prioritized over DHCPv6. So, it will take some more time. Once the clustering feature is implemented for DNS and server settings, the DHCP server is planned to have a major upgrade which will include making scopes more flexible, adding DHCPv6 support and adding failover support.

> Also, just curious on if you accept funding to prioritize features?

The project development is not my full time work, as its done after I get time from my day job. So its not feasible to commit time for some funding.

**Comment by josemanimala** [comment]
@ShreyasZare Thank you. 
Will be patiently waiting for the overhaul. IPv6 became more relevant recently with AWS starting to charge for IPv4 and most ISP networks starting to support it. I wanted to make sure I had my full home stack ready to switch to IPv6 when the time came.

**Comment by BazlTech** [comment]
Still holding out for DHCPv6!!

**Comment by BazlTech** [comment]
I know you have been very busy, and I don't mean to be a bother.  I'm just wondering if DHCP6 is on the roadmap yet.

**Comment by ShreyasZare** [comment]
> I know you have been very busy, and I don't mean to be a bother. I'm just wondering if DHCP6 is on the roadmap yet.

@BazlTech thanks for asking. DHCPv6 support along with major changes to existing DHCP server are planned but will be taken up after the clustering feature for the DNS server part is implemented. Clustering is being currently implemented in 2 stages after which DHCP upgrade is planned.

**Comment by BazlTech** [comment]
> > I know you have been very busy, and I don't mean to be a bother. I'm just wondering if DHCP6 is on the roadmap yet.
> 
> @BazlTech thanks for asking. DHCPv6 support along with major changes to existing DHCP server are planned but will be taken up after the clustering feature for the DNS server part is implemented. Clustering is being currently implemented in 2 stages after which DHCP upgrade is planned.

Is "Zone Catalogs" the same as clustering?  I see that was release and am just in anticipation of DHCPv6 support.  I do agree clustering would be more important than DHCPv6 though.

**Comment by ShreyasZare** [comment]
> > > I know you have been very busy, and I don't mean to be a bother. I'm just wondering if DHCP6 is on the roadmap yet.
> > 
> > 
> > @BazlTech thanks for asking. DHCPv6 support along with major changes to existing DHCP server are planned but will be taken up after the clustering feature for the DNS server part is implemented. Clustering is being currently implemented in 2 stages after which DHCP upgrade is planned.
> 
> Is "Zone Catalogs" the same as clustering? I see that was release and am just in anticipation of DHCPv6 support. I do agree clustering would be more important than DHCPv6 though.

Catalog zones provides "clustering" for only the Zones part on the admin panel. The clustering feature being designed will sync all settings, apps, etc. such that you can manage the entire cluster from a single node.

**Comment by stevapple** [comment]
> Option 82 seems to be too specialized that is supposed to be used by ISPs to communicate between the DHCP server and DHCP Relay Agents so does not seem to be useful for this project.

Add my scenario for option 82 here: we’re using an L3 switch with several DHCP pools configured for different VLANs. Now we would like to use Technitium to do DHCP as it can seamlessly add DNS records (crucial for DHCPv6). However, it’s not possible to physically connect the server to all these VLANs as a switch naturally does, so out solution is to use DHCP relay. Option 82, set by the switch based on VLAN ID, can then provide hints for which DHCP scope we want to use.

In short, I would love to see a feature that allows us to match option 82 (with RegEx?) and automatically picking up the desired scope to land alongside with DHCPv6.

**Comment by mentalinc** [comment]
How are people getting around having no DHCPv6, I don't think you can use Technitium for IPv4 and then Router for IPv6.

I've disabled DHCP on my router so I have a more stable 'authority' for IPv4 Static DHCP static IP. But without IPv6 at all, I'm going to need to revert to using my Router for DHCP

I have a static IP and a /56 allocated by the ISP if that makes any difference.

**Comment by gibbygano** [comment]
I just dropped pihole to switch to technitium and was very pleased until I realize that DHCPv6 isn't supported. None of my mobile devices will route to my local DNS entries even when I plug in IPv6 of the DNS server into my router's IPv6 settings. Am I missing something? It was that simple on pihole.

**Comment by BazlTech** [comment]
> How are people getting around having no DHCPv6, I don't think you can use Technitium for IPv4 and then Router for IPv6.
> 
> I've disabled DHCP on my router so I have a more stable 'authority' for IPv4 Static DHCP static IP. But without IPv6 at all, I'm going to need to revert to using my Router for DHCP
> 
> I have a static IP and a /56 allocated by the ISP if that makes any difference.

I'm using technitium for DHCP4.  I've got SLAAC setup on my L3 switch.  Works fine, aside from not being able to resolve IPv6 addresses locally.

Still looking forward to getting DHCP6 on technitium!

**Comment by rfpg1** [comment]
Since clustering is out, is DHCPv6 the next major release? Really looking forward to it

**Comment by ShreyasZare** [comment]
> Since clustering is out, is DHCPv6 the next major release? Really looking forward to it

Yes, but a few DNS features are planned before that and there may be a major release in between for .NET 10 runtime upgrade before the DHCPv6 release.

**Comment by kraduk** [comment]
can you let us know the timelines we should expect this release to be?

**Comment by ShreyasZare** [comment]
> can you let us know the timelines we should expect this release to be?

Its really not feasible to provide any timelines. Things are being added as and when time is available.

**Comment by realgeneraliroh** [comment]
I am just here to say I am glad it is on the "todo" list. 😄

**Comment by BazlTech** [comment]
Has this been assigned to an upcoming release stream/sprint yet?

**Comment by ShreyasZare** [comment]
> Has this been assigned to an upcoming release stream/sprint yet?

Not yet. This will be taken up after a couple of updates. A few DNS related features are being prioritized before taking this up. The DHCPv6, DHCPv4 major upgrade, and DHCP clustering will be taken up together and its going to take good amount of time to implement which is why a few DNS features are planned before so as to not delay them any further.

**Comment by mth9977** [comment]
Technitium is a very nice peace of software. congrats, you doa very good job.
But the lack of dhcpv6 support forces me to take another tool, very bad.
hope you'll find time to implement this .
kind regards from switzerland

**Comment by realgeneraliroh** [comment]
I know the pings are annoying, but I see the .NET runtime has been upgraded. Hopefully DHCPv6 is coming soon™.

**Comment by theotherguy2175** [comment]
Plus on this too
